File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pos/PosPreBuilt.java
package org.tdf.sunflower.consensus.pos;

import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.tdf.common.serialize.Codec;
import org.tdf.common.serialize.Codecs;
import org.tdf.common.store.PrefixStore;
import org.tdf.common.store.Store;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.ByteArrayMap;
import org.tdf.common.util.ByteUtil;
import org.tdf.common.util.HexBytes;
import org.tdf.common.util.RLPUtil;
import org.tdf.rlp.RLPCodec;
import org.tdf.rlp.RLPList;
import org.tdf.sunflower.state.Account;
import org.tdf.sunflower.state.AccountTrie;
import org.tdf.sunflower.state.Constants;
import org.tdf.sunflower.state.PreBuiltContract;
import org.tdf.sunflower.vm.Backend;
import org.tdf.sunflower.vm.CallData;
import org.tdf.sunflower.vm.abi.Abi;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Slf4j(topic = "pos")
public class PosPreBuilt implements PreBuiltContract {
    public static final HexBytes NODE_INFO_KEY = HexBytes.fromBytes("nodes".getBytes(StandardCharsets.US_ASCII));
    public static final HexBytes VOTE_INFO_KEY = HexBytes.fromBytes("votes".getBytes(StandardCharsets.US_ASCII));

    private final Map<HexBytes, NodeInfo> nodes;
    @Setter
    private AccountTrie accountTrie;

    public PosPreBuilt(@NonNull Map<HexBytes, NodeInfo> nodes) {
        this.nodes = nodes;
    }

    private static <T> Map.Entry<Integer, T> findFirst(List<? extends T> c, Predicate<T> predicate) {
        for (int i = 0; i < c.size(); i++) {
            if (predicate.test(c.get(i))) {
                return new AbstractMap.SimpleImmutableEntry<>(i, c.get(i));
            }
        }
        return new AbstractMap.SimpleImmutableEntry<>(-1, null);
    }

    private HexBytes getValue(HexBytes stateRoot, HexBytes key) {
        Account a = accountTrie.get(stateRoot, Constants.POS_CONTRACT_ADDR);
        Store<HexBytes, HexBytes> db = accountTrie.getContractStorageTrie().revert(a.getStorageRoot());
        return db.get(key);
    }

    public List<HexBytes> getNodes(HexBytes stateRoot) {
        HexBytes v = getValue(stateRoot, NODE_INFO_KEY);
        return Arrays.stream(RLPUtil.decode(v, NodeInfo[].class)).map(NodeInfo::getAddress).collect(Collectors.toList());
    }

    public List<NodeInfo> getNodeInfos(HexBytes stateRoot) {
        HexBytes v = getValue(stateRoot, NODE_INFO_KEY);
        return Arrays.asList(RLPUtil.decode(
                v, NodeInfo[].class));
    }

    public Optional<VoteInfo> getVoteInfo(HexBytes stateRoot, HexBytes txHash) {
        Account a = accountTrie.get(stateRoot, Constants.POS_CONTRACT_ADDR);
        Store<HexBytes, HexBytes> db = accountTrie.getContractStorageTrie().revert(a.getStorageRoot());
        PrefixStore<HexBytes, VoteInfo> store = getVoteInfoStore(db);
        return Optional.ofNullable(store.get(txHash));
    }

    public PrefixStore<HexBytes, VoteInfo> getVoteInfoStore(Store<HexBytes, HexBytes> contractStore) {
        return new PrefixStore<>(
                contractStore,
                VOTE_INFO_KEY,
                Codecs.HEX,
                Codecs.newRLPCodec(VoteInfo.class)
        );
    }

    @Override
    public Account getGenesisAccount() {
        return Account.emptyAccount(Constants.POS_CONTRACT_ADDR, Uint256.ZERO);
    }

    @Override
    public Map<HexBytes, HexBytes> getGenesisStorage() {
        Map<HexBytes, HexBytes> map = new HashMap<>();
        List<NodeInfo> nodeInfos = new ArrayList<>(this.nodes.values());
        nodeInfos.sort(NodeInfo::compareTo);
        map.put(NODE_INFO_KEY, RLPUtil.encode(nodeInfos));
        map.put(VOTE_INFO_KEY, RLPUtil.encode(RLPList.createEmpty()));
        return map;
    }

    @Override
    @SneakyThrows
    public byte[] call(Backend backend, CallData callData) {
        HexBytes payload = callData.getData();
        Type type = Type.values()[payload.get(0)];
        HexBytes args = payload.slice(1);

        List<NodeInfo> nodeInfos = new ArrayList<>(
                Arrays.asList(
                        RLPUtil.decode(backend.dbGet(Constants.POS_CONTRACT_ADDR, NODE_INFO_KEY), NodeInfo[].class))
        );

        PrefixStore<HexBytes, VoteInfo> voteInfos = getVoteInfoStore(backend.getAsStore(Constants.POS_CONTRACT_ADDR));

        switch (type) {
            case VOTE: {
                if (callData.getValue().compareTo(Uint256.ZERO) == 0)
                    throw new RuntimeException("amount of vote cannot be 0 ");
                Map.Entry<Integer, NodeInfo> e =
                        findFirst(nodeInfos, x -> x.address.equals(args));

                NodeInfo n = e.getKey() < 0 ? new NodeInfo(args, Uint256.ZERO, new ArrayList<>()) :
                        e.getValue();

                n.vote = n.vote.safeAdd(callData.getValue());
                n.txHash.add(callData.getTxHash());
                if (e.getKey() < 0)
                    nodeInfos.add(n);
                else
                    nodeInfos.set(e.getKey(), n);
                voteInfos.put(callData.getTxHash(), new VoteInfo(
                        callData.getTxHash(), callData.getCaller(),
                        args, callData.getValue()
                ));
                break;
            }
            case CANCEL_VOTE:
                if (callData.getValue().compareTo(Uint256.ZERO) != 0)
                    throw new RuntimeException("amount of cancel vote should be 0");
                Optional<VoteInfo> o = Optional.ofNullable(voteInfos.get(args));
                voteInfos.remove(args);

                if (!o.isPresent()) {
                    throw new RuntimeException(args + " voting business does not exist and cannot be withdrawn");
                }
                VoteInfo voteInfo = o.get();
                if (!voteInfo.getFrom().equals(callData.getCaller())) {
                    throw new RuntimeException("vote transaction from " + voteInfo.getFrom() + " not equals to " + callData.getCaller());
                }

                Map.Entry<Integer, NodeInfo> e2 =
                        findFirst(nodeInfos, x -> x.address.equals(voteInfo.to));

                if (e2.getKey() < 0) {
                    throw new RuntimeException(voteInfo.getTo() + " abnormal withdrawal of vote");
                }

                NodeInfo ninfo = e2.getValue();

                if (!ninfo.txHash.contains(args))
                    throw new RuntimeException("vote " + args + " not exists");
                ninfo.vote = ninfo.vote.safeSub(voteInfo.amount);
                ninfo.txHash.remove(args);

                if (ninfo.vote.compareTo(Uint256.ZERO) == 0) {
                    nodeInfos.remove((int) e2.getKey());
                } else {
                    nodeInfos.set(e2.getKey(), ninfo);
                }


                Uint256 callerBalance = backend.getBalance(callData.getCaller());
                backend.setBalance(callData.getCaller(), callerBalance.add(voteInfo.amount));

                Uint256 thisBalance = backend.getBalance(Constants.POS_CONTRACT_ADDR);
                backend.setBalance(Constants.POS_CONTRACT_ADDR, thisBalance.safeSub(voteInfo.amount));
                break;
        }
        nodeInfos.sort(NodeInfo::compareTo);
        Collections.reverse(nodeInfos);
        backend.dbSet(Constants.POS_CONTRACT_ADDR, NODE_INFO_KEY, RLPUtil.encode(nodeInfos));
        return ByteUtil.EMPTY_BYTE_ARRAY;
    }

    @Override
    public Abi getAbi() {
        return Abi.fromJson("[]");
    }

    public enum Type {
        VOTE,
        CANCEL_VOTE
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class NodeInfo implements Comparable<NodeInfo> {
        private HexBytes address;
        private Uint256 vote;
        private List<HexBytes> txHash;

        @Override
        public int compareTo(NodeInfo o) {
            return vote.compareTo(o.vote);
        }
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class VoteInfo implements Comparable<VoteInfo> {
        private HexBytes txHash;
        private HexBytes from;
        private HexBytes to;
        private Uint256 amount;

        @Override
        public int compareTo(VoteInfo o) {
            return txHash.compareTo(o.txHash);
        }
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/consensus/pos/PosPreBuilt.kt
package org.tdf.sunflower.consensus.pos

import com.github.salpadding.rlpstream.Rlp
import com.github.salpadding.rlpstream.annotation.RlpCreator
import com.github.salpadding.rlpstream.annotation.RlpProps
import org.slf4j.LoggerFactory
import org.tdf.common.serialize.Codecs
import org.tdf.common.serialize.Codecs.newRLPCodec
import org.tdf.common.store.PrefixStore
import org.tdf.common.store.Store
import org.tdf.common.types.Uint256
import org.tdf.common.util.ByteUtil
import org.tdf.common.util.HexBytes
import org.tdf.common.util.RLPUtil
import org.tdf.sunflower.facade.RepositoryReader
import org.tdf.sunflower.state.AccountTrie
import org.tdf.sunflower.state.BuiltinContract
import org.tdf.sunflower.state.Constants
import org.tdf.sunflower.vm.Backend
import org.tdf.sunflower.vm.CallContext
import org.tdf.sunflower.vm.CallData
import org.tdf.sunflower.vm.abi.Abi
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.AbstractMap.SimpleImmutableEntry
import java.util.function.Predicate

class PosPreBuilt(private val nodes: Map<HexBytes, NodeInfo>) : BuiltinContract {
    lateinit var accountTrie: AccountTrie

    private fun getValue(stateRoot: HexBytes, key: HexBytes): HexBytes? {
        val a = accountTrie.get(stateRoot, Constants.POS_CONTRACT_ADDR)!!
        val db: Store<HexBytes, HexBytes> = accountTrie.contractStorageTrie.revert(a.storageRoot)
        return db[key]
    }

    fun getNodes(stateRoot: HexBytes): List<HexBytes> {
        val v = getValue(stateRoot, NODE_INFO_KEY)!!
        return RLPUtil.decode(v, Array<NodeInfo>::class.java).map { it.address }
    }

    fun getNodeInfos(stateRoot: HexBytes): List<NodeInfo> {
        val v = getValue(stateRoot, NODE_INFO_KEY)
        return RLPUtil.decode(v!!, Array<NodeInfo>::class.java).toList()

    }

    fun getVoteInfo(stateRoot: HexBytes, txHash: HexBytes): VoteInfo? {
        val a = accountTrie.get(stateRoot, Constants.POS_CONTRACT_ADDR)!!
        val db: Store<HexBytes, HexBytes> = accountTrie.contractStorageTrie.revert(a.storageRoot)
        val store = getVoteInfoStore(db)
        return store[txHash]
    }

    private fun getVoteInfoStore(contractStore: Store<HexBytes, HexBytes>): PrefixStore<HexBytes, VoteInfo> {
        return PrefixStore(
            contractStore,
            VOTE_INFO_KEY,
            Codecs.HEX,
            newRLPCodec(VoteInfo::class.java)
        )
    }

    override val address: HexBytes
        get() = Constants.POS_CONTRACT_ADDR
    override val genesisStorage: Map<HexBytes, HexBytes>
        get() {
            val map: MutableMap<HexBytes, HexBytes> = HashMap()
            val nodeInfos = nodes.values.toMutableList()
            nodeInfos.sortWith{ obj: NodeInfo, o: NodeInfo -> obj.compareTo(o) }
            map[NODE_INFO_KEY] = RLPUtil.encode(nodeInfos)
            map[VOTE_INFO_KEY] = HexBytes.fromBytes(Rlp.encodeElements())
            return map
        }

    override fun call(rd: RepositoryReader, backend: Backend, ctx: CallContext, callData: CallData): ByteArray {
        val payload = callData.data
        val type = Type.values()[payload[0]]
        val args = payload.slice(1)
        val nodeInfos: MutableList<NodeInfo> =
            RLPUtil.decode(backend.dbGet(Constants.POS_CONTRACT_ADDR, NODE_INFO_KEY), Array<NodeInfo>::class.java)
                .toMutableList()


        val voteInfos = getVoteInfoStore(backend.getAsStore(Constants.POS_CONTRACT_ADDR))
        when (type) {
            Type.VOTE -> {
                if (callData.value.compareTo(Uint256.ZERO) == 0) throw RuntimeException("amount of vote cannot be 0 ")
                val (key, value) = findFirst(nodeInfos) { it.address == args }
                val n = if (key < 0) NodeInfo(args, Uint256.ZERO, ArrayList()) else value!!
                val n1 = n.copy(vote = n.vote + callData.value, txHash = n.txHash + ctx.txHash)
                if (key < 0) nodeInfos.add(n1) else nodeInfos[key] = n1
                voteInfos[ctx.txHash] = VoteInfo(
                    ctx.txHash, callData.caller,
                    args, callData.value
                )
            }
            Type.CANCEL_VOTE -> {
                if (callData.value != Uint256.ZERO)
                    throw RuntimeException("amount of cancel vote should be 0")

                val voteInfo = voteInfos[args] ?: throw RuntimeException("$args voting business does not exist and cannot be withdrawn")
                voteInfos.remove(args)
                if (voteInfo.from != callData.caller) {
                    throw RuntimeException("vote transaction from " + voteInfo.from + " not equals to " + callData.caller)
                }
                val (key, ninfo) = findFirst(nodeInfos) { it.address == voteInfo.to }
                if (ninfo == null) {
                    throw RuntimeException(voteInfo.to.toString() + " abnormal withdrawal of vote")
                }
                if (!ninfo.txHash.contains(args)) throw RuntimeException("vote $args not exists")
                val n1 = ninfo.copy(vote = ninfo.vote - voteInfo.amount, txHash = ninfo.txHash - args)
                if (ninfo.vote == Uint256.ZERO) {
                    nodeInfos.removeAt(key)
                } else {
                    nodeInfos[key] = n1
                }
                val callerBalance = backend.getBalance(callData.caller)
                backend.setBalance(callData.caller, callerBalance.plus(voteInfo.amount))
                val thisBalance = backend.getBalance(Constants.POS_CONTRACT_ADDR)
                backend.setBalance(Constants.POS_CONTRACT_ADDR, thisBalance.minus(voteInfo.amount))
            }
        }
        nodeInfos.sortWith { obj: NodeInfo, o: NodeInfo -> obj.compareTo(o) }
        nodeInfos.reverse()
        backend.dbSet(Constants.POS_CONTRACT_ADDR, NODE_INFO_KEY, RLPUtil.encode(nodeInfos))
        return ByteUtil.EMPTY_BYTE_ARRAY
    }

    override val abi: Abi
        get() = Abi.fromJson("[]")

    enum class Type {
        VOTE, CANCEL_VOTE
    }

    @RlpProps("address", "vote", "txHash")
    data class NodeInfo @RlpCreator constructor(
        val address: HexBytes = HexBytes.empty(),
        val vote: Uint256 = Uint256.ZERO,
        val txHash: List<HexBytes> = emptyList()
    ): Comparable<NodeInfo> {
        override fun compareTo(other: NodeInfo): Int {
            return vote.compareTo(other.vote)
        }
    }


    @RlpProps("txHash", "from", "to", "amount")
    class VoteInfo @RlpCreator constructor(
        val txHash: HexBytes = HexBytes.empty(),
        val from: HexBytes = HexBytes.empty(),
        val to: HexBytes = HexBytes.empty(),
        val amount: Uint256 = Uint256.ZERO
    ): Comparable<VoteInfo> {
        override fun compareTo(other: VoteInfo): Int {
            return txHash.compareTo(other.txHash)
        }
    }

    companion object {
        val NODE_INFO_KEY = "nodes".ascii()
        val VOTE_INFO_KEY = "votes".ascii()
        private val log = LoggerFactory.getLogger("pos")

        private fun <T> findFirst(c: List<T>, predicate: Predicate<T>): Map.Entry<Int, T?> {
            for (i in c.indices) {
                if (predicate.test(c[i])) {
                    return SimpleImmutableEntry(i, c[i])
                }
            }
            return SimpleImmutableEntry<Int, T?>(-1, null)
        }
    }
}


internal fun String.ascii(): HexBytes {
    return HexBytes.fromBytes(this.toByteArray(StandardCharsets.US_ASCII))
}
